pkg/controller/replicaschedulingpreference/controller.go

```diff
@@ -0,0 +1,751 @@
+			Group:      "apps",
+			Version:    "v1",
+			Kind:       "deployment",
+			Namespaced: true,
+		},
+		func(obj pkgruntime.Object) {
+			s.deliverObj(obj, s.reviewDelay, false)
+		},
+		&util.ClusterLifecycleHandlerFuncs{
+			ClusterAvailable: func(cluster *fedv1a1.FederatedCluster) {
+				// When new cluster becomes available process all the target resources again.
+				s.clusterDeliverer.DeliverAt(allClustersKey, nil, time.Now().Add(s.clusterAvailableDelay))
+			},
+			// When a cluster becomes unavailable process all the target resources again.
+			ClusterUnavailable: func(cluster *fedv1a1.FederatedCluster, _ []interface{}) {
+				s.clusterDeliverer.DeliverAt(allClustersKey, nil, time.Now().Add(s.clusterUnavailableDelay))
+			},
+		},
+	)
+
+	// Federated informer watching pods in members of federation.
+	s.podInformer = util.NewFederatedInformer(
+		fedClient,
+		kubeClient,
+		crClient,
+		&metav1.APIResource{
+			Name:       "pods",
+			Group:      "",
+			Version:    "v1",
+			Kind:       "pod",
+			Namespaced: true,
+		},
+		func(obj pkgruntime.Object) {
+			s.deliverObj(obj, s.reviewDelay, false)
+		},
+		&util.ClusterLifecycleHandlerFuncs{
+			ClusterAvailable: func(cluster *fedv1a1.FederatedCluster) {
+				// When new cluster becomes available process all the target resources again.
+				s.clusterDeliverer.DeliverAt(allClustersKey, nil, time.Now().Add(s.clusterAvailableDelay))
+			},
+			// When a cluster becomes unavailable process all the target resources again.
+			ClusterUnavailable: func(cluster *fedv1a1.FederatedCluster, _ []interface{}) {
+				s.clusterDeliverer.DeliverAt(allClustersKey, nil, time.Now().Add(s.clusterUnavailableDelay))
+			},
+		},
+	)
+
+	return s, nil
+}
+
+// minimizeLatency reduces delays and timeouts to make the controller more responsive (useful for testing).
+func (s *ReplicaSchedulingPreferenceController) minimizeLatency() {
+	s.clusterAvailableDelay = time.Second
+	s.clusterUnavailableDelay = time.Second
+	s.reviewDelay = 50 * time.Millisecond
+	s.smallDelay = 20 * time.Millisecond
+	s.updateTimeout = 5 * time.Second
+}
+
+func (s *ReplicaSchedulingPreferenceController) Run(stopChan <-chan struct{}) {
+	go s.controller.Run(stopChan)
+	go s.templateController.Run(stopChan)
+	go s.overrideController.Run(stopChan)
+	go s.placementController.Run(stopChan)
+
+	s.targetInformer.Start()
+	s.podInformer.Start()
+	s.deliverer.StartWithHandler(func(item *util.DelayingDelivererItem) {
+		s.workQueue.Add(item)
+	})
+	s.clusterDeliverer.StartWithHandler(func(_ *util.DelayingDelivererItem) {
+		s.reconcileOnClusterChange()
+	})
+
+	go wait.Until(s.worker, time.Second, stopChan)
+
+	util.StartBackoffGC(s.backoff, stopChan)
+
+	// Ensure all goroutines are cleaned up when the stop channel closes
+	go func() {
+		<-stopChan
+		s.targetInformer.Stop()
+		s.podInformer.Stop()
+		s.workQueue.ShutDown()
+		s.deliverer.Stop()
+		s.clusterDeliverer.Stop()
+	}()
+}
+
+type reconciliationStatus int
+
+const (
```

Comment:
(No action required) Seems like there is common controller code that could be factored out for reuse at some point.

Reply:
Filed https://github.com/kubernetes-sigs/federation-v2/issues/102